whc-shared/src/main/java/com/ibm/whc/deid/shared/pojo/config/masking/BinningMaskingProviderConfig.java

```diff
@@ -17,99 +17,169 @@
+		return useSingleBucketOverThreshold;
+	}
+
+	public void setUseSingleBucketOverThreshold(boolean useSingleBucketOverThreshold) {
+		this.useSingleBucketOverThreshold = useSingleBucketOverThreshold;
+	}
+
+	public double getSingleBucketOverThresholdValue() {
+		return singleBucketOverThresholdValue;
+	}
+
+	public void setSingleBucketOverThresholdValue(double singleBucketOverThresholdValue) {
+		this.singleBucketOverThresholdValue = singleBucketOverThresholdValue;
+	}
+
+	public String getSingleBucketOverThresholdReplacement() {
+		return singleBucketOverThresholdReplacement;
+	}
+
+	public void setSingleBucketOverThresholdReplacement(String singleBucketOverThresholdReplacement) {
+		this.singleBucketOverThresholdReplacement = singleBucketOverThresholdReplacement;
+	}
+
+	public boolean isUseSingleBucketUnderThreshold() {
+		return useSingleBucketUnderThreshold;
+	}
+
+	public void setUseSingleBucketUnderThreshold(boolean useSingleBucketUnderThreshold) {
+		this.useSingleBucketUnderThreshold = useSingleBucketUnderThreshold;
+	}
+
+	public double getSingleBucketUnderThresholdValue() {
+		return singleBucketUnderThresholdValue;
+	}
+
+	public void setSingleBucketUnderThresholdValue(double singleBucketUnderThresholdValue) {
+		this.singleBucketUnderThresholdValue = singleBucketUnderThresholdValue;
+	}
+
+	public String getSingleBucketUnderThresholdReplacement() {
+		return singleBucketUnderThresholdReplacement;
+	}
+
+	public void setSingleBucketUnderThresholdReplacement(String singleBucketUnderThresholdReplacement) {
+		this.singleBucketUnderThresholdReplacement = singleBucketUnderThresholdReplacement;
+	}
+	
+	@Override
+	public void validate(DeidMaskingConfig maskingConfig) throws InvalidMaskingConfigurationException {
+		super.validate(maskingConfig);
+		if (binSize < 1) {
+			throw new InvalidMaskingConfigurationException("`binSize` must be greater than 0");
+		}
+		 if (singleBucketOverThresholdReplacement == null) {
+		      throw new InvalidMaskingConfigurationException("`singleBucketOverThresholdReplacement` must not be null");
+		 }
+		 if (singleBucketUnderThresholdReplacement == null) {
+		      throw new InvalidMaskingConfigurationException("`singleBucketUnderThresholdReplacement` must not be null");
+		 }
+		 if (singleBucketOverThresholdValue < 1) {
+				throw new InvalidMaskingConfigurationException("`singleBucketOverThresholdValue` must be greater than 0");
+			}
+		 if (singleBucketUnderThresholdValue < 1) {
+				throw new InvalidMaskingConfigurationException("`singleBucketUnderThresholdValue` must be greater than 0");
+			}
+		if (useSingleBucketOverThreshold == true && useSingleBucketUnderThreshold == true) {
+			if (singleBucketOverThresholdValue < singleBucketUnderThresholdValue) {
+				throw new InvalidMaskingConfigurationException("`singleBucketOverThresholdValue` must be greater than singleBucketUnderThresholdValue");
+			}
+		}
+	}
+
+	@Override
+	public int hashCode() {

```

Comment:
A few notes on equals() and hashCode() - Each class is allowed to decide what has to be equal for an instance of the class to be considered equal to a different object.  For this class, we want each instance to only be equal to another instance of the same type and we want every field to be the same in both instances.  For other classes, equal() can be implemented totally differently.

The hashCode() method is called when an instance is inserted as a key into a construct like a HashSet or a HashMap.  There could be other reasons, but that is the most common.  You want to generate a code that is reasonably unique given the values contained in the object.  It is also generally required, although it is not enforced, that two instances return the same value from hashCode() if equals() for the same two instances would be true.  Since we want to consider every field as part of equals(), we need to consider every field when we compute the hashCode().

Writing equals() and hashCode() methods that consider every field is VERY tedious and the testing of these methods even more so.  Fortunately, eclipse has a built-in function to generate these methods for us now, which is what I use for any new or updated code.  Just remove hashCode() and equals(), right-click on the part, and use the option under Generate to generate the equals() and hashCode() methods.  I usually take all the options the generation offers except the one to generate comments - but you can add comments too if you would like.

Since the methods are generated, I typically do not generate test cases for them.  I have done that in the past to make sure the generation is working well, but it is extremely tedious and generally not worth the time.  I believe the algorithm is solid and a visual examination of what it produced is sufficient to make sure our input wasn't incorrect.